### ISSUE TITLE
InternalServerError: topological ref error

### ISSUE THREAD

**Issue by aljazerzen** [edgedb/edgedb#6525]
Steps to reproduce:

1. `edb cli migration create`

Schema:
```
module default {

    global currentUserId: uuid;
    global currentUser:= (
        SELECT User FILTER .userId = global currentUserId AND .status = Status.ACTIVE LIMIT 1
    );
    global isCurrentUser:= (
        exists global currentUser
    );
    global currentTenant:= (
        global currentUser.tenant
    );
    global currentOrganization:= (
        global currentUser.tenant.organization
    );
    global superUser:= (
        SELECT SuperUser FILTER .userId = global currentUserId AND .status = Status.ACTIVE LIMIT 1
    );
    global isSuperUser:= (
        exists global superUser or (SELECT count(SuperUser) = 0)
    );
    global isTenantAdmin:= (
        UserRole.TENANT_ADMIN in global currentUser.roles
    );
    global isManager:= (
        UserRole.MANAGER in global currentUser.roles
    );
    global isLoanOfficer:= (
        UserRole.LOAN_OFFICER in global currentUser.roles
    );
    global isPOS: bool;

    scalar type Status extending enum<ACTIVE, BANNED>;
    scalar type UserRole extending enum<Normal, Admin>;

    type User {
        status: Status;
        role: UserRole;

        tenant: Tenant;
    };

    type Tenant {
        organization: Organization;
    };

    type Organization {
        currencies: array<str>;
    };

    type SuperUser extending User {
    };

    type ProfileAccount {
        name: str;
        position: int16;
        status: Status;
        profile: Profile;    
        currency: str;
    }

    scalar type ProfileType extending enum<PERSON, COMPANY, GROUP_MEMBER>;

    scalar type ProfileStatus extending enum<Online, Offline>;

    type Profile {
        annotation description:= 'Profiles: Profile: Primary data';

        required name: str;
        required code: str {
            readonly:= true;
        };
        required profileType: ProfileType {
            readonly:= true;
            default:= ProfileType.PERSON;
        };
        required profileStatus: ProfileStatus;

        index pg::btree on ((.tenant, .profileType, .profileStatus, .modifiedAt));
        index pg::btree on ((.tenant, .profileType, .modifiedAt));

        trigger create_profileAccount after insert for each do (
            WITH
                currenciesList:= (global currentOrganization.currencies)
            FOR currencyItem IN currenciesList UNION (
                INSERT ProfileAccount {name:= 'Current account: ' ++ currencyItem.code, position:= <int16>round(random() * 100),
                status:= Status.ACTIVE, profile:= <Profile>__new__.id, currency:= currencyItem})
        );

        #InternalServerError: there is no range var for ns~34@@(__derived__::__new__).>sourceEntity[IS default::FieldDataSource] source in <pg.SelectStmt at 0x7fdbc9fd4c90>
        #access policy superUser_full_access allow all
        #    using (global isSuperUser ?? false);
        #access policy tenantAdmin_full_access allow all
        #    using (global currentTenant ?= .tenant and global isTenantAdmin);
        #access policy manager_officer_insert_update allow insert, update
        #    using (global currentTenant ?= .tenant and (global isManager or global isLoanOfficer));
        #access policy user_select allow select
        #    using (global currentTenant ?= .tenant);
    }

    type CustomField {
        link customSection: CustomSection;
    };

    type CustomSection {
        customType: CustomType;
    };

    scalar type CustomType extending enum<LOAN_APP, PERSON, COMPANY, GROUP_MEMBER>;

    type FieldDataSource {
    };

    abstract type CustomData {
        annotation description:= 'Profile and Loan application CustomField meta-data';

        property customType:= .customField.customSection.customType;

        required customField: CustomField {
            readonly:= true;
        };
        required sourceEntity: FieldDataSource {
            readonly:= true;
            on target delete delete source;
        };
        link customSection:= .customField.customSection;

        constraint exclusive on ((.sourceEntity, .customField));

        trigger check_customType after insert for each do (
            assert(__new__.customType = CustomType.LOAN_APP if __new__.sourceEntity is LoanApp else true,
                message:= 'The source object must be a loan application') and
            assert(
                (WITH
                    profile:= (<Profile>__new__.sourceEntity.id)
                SELECT (__new__.customType = CustomType.PERSON and profile.profileType = ProfileType.PERSON) or
                    (__new__.customType = CustomType.COMPANY and profile.profileType = ProfileType.COMPANY) or
                    (__new__.customType = CustomType.GROUP_MEMBER and profile.profileType = ProfileType.GROUP_MEMBER))
                if __new__.sourceEntity is Profile else true, message:= 'The source profile object has a wrong type')
        );

        access policy superUser_full_access allow all
            using (global isSuperUser ?? false);
        access policy tenantAdmin_full_access allow all
            using (global currentTenant ?= .tenant and global isTenantAdmin);
        access policy manager_officer_insert_update allow insert, update
            using (global currentTenant ?= .tenant and (global isManager or global isLoanOfficer));
        access policy pos_insert allow insert
            using (global currentTenant ?= .tenant and global isPOS);
        access policy user_select allow select
            using (global currentTenant ?= .tenant);
    };

    type X extending CustomData {
        
    };    
}
```

**Comment by mkhlsavin** [comment]
@aljazerzen I'll send full schema today by Discord

**Comment by mkhlsavin** [comment]
To reproduce this error, you need to create a new database instance and try to compile the schema. There is about a 50% chance you will receive one of the following errors:

1. SchemaDefinitionError: cannot alter object type 'default::User' because this affects body expression of function 'func::getAccrualStatus(id: std::uuid, date: cal::local_date)'  Detail: cannot alter function 'func::getAccrualStatus(id: std::uuid, date: cal::local_date)' because this affects expression of trigger 'update_loanAccruals' of object type 'default::LoanPayments'
2. SchemaDefinitionError: cannot alter object type 'default::LoanPayments' because this affects body expression of function 'func::getLateDays(id: std::uuid, date: cal::local_date)'  Detail: cannot alter function 'func::getLateDays(id: std::uuid, date: cal::local_date)' because this affects expression of trigger 'update_loanAccruals' of object type 'default::LoanPayments'
3. SchemaDefinitionError: cannot alter object type 'default::LoanPayments' because this affects body expression of function 'func::getLastLoanPayment(loan: default::Loan, date: cal::local_date)'  Detail: cannot alter function 'func::getLastLoanPayment(loan: default::Loan, date: cal::local_date)' because this affects expression of trigger 'update_loanAccruals' of object type 'default::LoanPayments'
4. SchemaDefinitionError: cannot alter object type 'default::LoanPayments' because this affects body expression of function 'func::getOutstandingBalance(loan: default::Loan, loanPayment: OPTIONAL default::LoanPayments)'  Detail: cannot alter function 'func::getOutstandingBalance(loan: default::Loan, loanPayment: OPTIONAL default::LoanPayments)' because this affects expression of trigger 'update_loanAccruals' of object type 'default::LoanPayments'

**Log with an example of schema compilation attempts:**

```
PS C:\Users\msavin\Documents\edb> edgedb project init
Found `edgedb.toml` in \\?\C:\Users\msavin\Documents\edb
Initializing project...
Specify the name of EdgeDB instance to use with this project [default: edb]:
> edb
Checking EdgeDB versions...
┌────────────────────────┬───────────────────────────────────────────────┐
│ Project directory      │ \\?\C:\Users\msavin\Documents\edb             │
│ Project config         │ \\?\C:\Users\msavin\Documents\edb\edgedb.toml │
│ Schema dir (non-empty) │ \\?\C:\Users\msavin\Documents\edb\dbschema    │
│ Installation method    │ WSL                                           │
│ Version                │ 4.1+2194ddc                                   │
│ Instance name          │ edb                                           │
└────────────────────────┴───────────────────────────────────────────────┘
Version 4.1+2194ddc is already downloaded
Initializing EdgeDB instance...
Applying migrations...
Everything is up to date. Revision initial
Project initialized.
To connect to edb, run `edgedb`
PS C:\Users\msavin\Documents\edb> edgedb migration create
Connecting to EdgeDB instance 'edb' at localhost:10700...
edgedb error: SchemaDefinitionError: cannot alter object type 'default::LoanPayments' because this affects body expression of function 'func::getLastLoanPayment(loan: default::Loan, date: cal::local_date)'
  Detail: cannot alter function 'func::getLastLoanPayment(loan: default::Loan, date: cal::local_date)' because this affects expression of trigger 'update_loanAccruals' of object type 'default::LoanPayments'
edgedb error: cannot proceed until .esdl files are fixed
PS C:\Users\msavin\Documents\edb> edgedb instance destroy -I edb --force
Unlinking \\?\C:\Users\msavin\Documents\edb
Instance edb is successfully deleted.
PS C:\Users\msavin\Documents\edb> edgedb project init
Found `edgedb.toml` in \\?\C:\Users\msavin\Documents\edb
Initializing project...
Specify the name of EdgeDB instance to use with this project [default: edb]:
> edb
Checking EdgeDB versions...
┌────────────────────────┬───────────────────────────────────────────────┐
│ Project directory      │ \\?\C:\Users\msavin\Documents\edb             │
│ Project config         │ \\?\C:\Users\msavin\Documents\edb\edgedb.toml │
│ Schema dir (non-empty) │ \\?\C:\Users\msavin\Documents\edb\dbschema    │
│ Installation method    │ WSL                                           │
│ Version                │ 4.1+2194ddc                                   │
│ Instance name          │ edb                                           │
└────────────────────────┴───────────────────────────────────────────────┘

Version 4.1+2194ddc is already downloaded
Initializing EdgeDB instance...
Applying migrations...
Everything is up to date. Revision initial
Project initialized.
To connect to edb, run `edgedb`
PS C:\Users\msavin\Documents\edb> edgedb migration create
Connecting to EdgeDB instance 'edb' at localhost:10700...
edgedb error: SchemaDefinitionError: cannot alter object type 'default::LoanPayments' because this affects body expression of function 'func::getOutstandingBalance(loan: default::Loan, loanPayment: OPTIONAL default::LoanPayments)'
  Detail: cannot alter function 'func::getOutstandingBalance(loan: default::Loan, loanPayment: OPTIONAL default::LoanPayments)' because this affects expression of trigger 'update_loanAccruals' of object type 'default::LoanPayments'
edgedb error: cannot proceed until .esdl files are fixed
PS C:\Users\msavin\Documents\edb> edgedb migration create
Connecting to EdgeDB instance 'edb' at localhost:10700...
edgedb error: SchemaDefinitionError: cannot alter object type 'default::LoanPayments' because this affects body expression of function 'func::getOutstandingBalance(loan: default::Loan, loanPayment: OPTIONAL default::LoanPayments)'
  Detail: cannot alter function 'func::getOutstandingBalance(loan: default::Loan, loanPayment: OPTIONAL default::LoanPayments)' because this affects expression of trigger 'update_loanAccruals' of object type 'default::LoanPayments'
edgedb error: cannot proceed until .esdl files are fixed
PS C:\Users\msavin\Documents\edb> edgedb instance destroy -I edb --force
Unlinking \\?\C:\Users\msavin\Documents\edb
Instance edb is successfully deleted.
PS C:\Users\msavin\Documents\edb> edgedb project init
Found `edgedb.toml` in \\?\C:\Users\msavin\Documents\edb
Initializing project...
Specify the name of EdgeDB instance to use with this project [default: edb]:
> edb
Checking EdgeDB versions...
┌────────────────────────┬───────────────────────────────────────────────┐
│ Project directory      │ \\?\C:\Users\msavin\Documents\edb             │
│ Project config         │ \\?\C:\Users\msavin\Documents\edb\edgedb.toml │
│ Schema dir (non-empty) │ \\?\C:\Users\msavin\Documents\edb\dbschema    │
│ Installation method    │ WSL                                           │
│ Version                │ 4.1+2194ddc                                   │
│ Instance name          │ edb                                           │
└────────────────────────┴───────────────────────────────────────────────┘
Version 4.1+2194ddc is already downloaded
Initializing EdgeDB instance...
Applying migrations...
Everything is up to date. Revision initial
Project initialized.
To connect to edb, run `edgedb`
PS C:\Users\msavin\Documents\edb> edgedb migration create
Connecting to EdgeDB instance 'edb' at localhost:10700...
edgedb error: SchemaDefinitionError: cannot alter object type 'default::LoanPayments' because this affects body expression of function 'func::getLateDays(id: std::uuid, date: cal::local_date)'
  Detail: cannot alter function 'func::getLateDays(id: std::uuid, date: cal::local_date)' because this affects expression of trigger 'update_loanAccruals' of object type 'default::LoanPayments'
edgedb error: cannot proceed until .esdl files are fixed
PS C:\Users\msavin\Documents\edb> edgedb instance destroy -I edb --force
Unlinking \\?\C:\Users\msavin\Documents\edb
Instance edb is successfully deleted.
PS C:\Users\msavin\Documents\edb> edgedb project init
Found `edgedb.toml` in \\?\C:\Users\msavin\Documents\edb
Initializing project...
Specify the name of EdgeDB instance to use with this project [default: edb]:
> edb
Checking EdgeDB versions...
┌────────────────────────┬───────────────────────────────────────────────┐
│ Project directory      │ \\?\C:\Users\msavin\Documents\edb             │
│ Project config         │ \\?\C:\Users\msavin\Documents\edb\edgedb.toml │
│ Schema dir (non-empty) │ \\?\C:\Users\msavin\Documents\edb\dbschema    │
│ Installation method    │ WSL                                           │
│ Version                │ 4.1+2194ddc                                   │
│ Instance name          │ edb                                           │
└────────────────────────┴───────────────────────────────────────────────┘
Version 4.1+2194ddc is already downloaded
Initializing EdgeDB instance...
Applying migrations...
Everything is up to date. Revision initial
Project initialized.
To connect to edb, run `edgedb`
PS C:\Users\msavin\Documents\edb> edgedb migration create
Connecting to EdgeDB instance 'edb' at localhost:10700...
Created C:\Users\msavin\Documents\edb\dbschema\migrations\00001.edgeql, id: m1xcr6kicdlddpgc7y4wpwfvgwqeqkkmy56ssn3m5onxlaug53zdga
```

**And same on 4.2 release:**

```
PS C:\Users\msavi\Documents\melfyn-v0\edb> edgedb project init
Found `edgedb.toml` in \\?\C:\Users\msavi\Documents\melfyn-v0\edb
Initializing project...
Specify the name of EdgeDB instance to use with this project [default: edb]:
> edb
Checking EdgeDB versions...
┌────────────────────────┬────────────────────────────────────────────────────────┐
│ Project directory      │ \\?\C:\Users\msavi\Documents\melfyn-v0\edb             │
│ Project config         │ \\?\C:\Users\msavi\Documents\melfyn-v0\edb\edgedb.toml │
│ Schema dir (non-empty) │ \\?\C:\Users\msavi\Documents\melfyn-v0\edb\dbschema    │
│ Installation method    │ WSL                                                    │
│ Version                │ 4.2+8778fc1                                            │
│ Instance name          │ edb                                                    │
└────────────────────────┴────────────────────────────────────────────────────────┘
Version 4.2+8778fc1 is already downloaded
Initializing EdgeDB instance...
Applying migrations...
Everything is up to date. Revision initial
Project initialized.
To connect to edb, run `edgedb`
PS C:\Users\msavi\Documents\melfyn-v0\edb> edgedb migration create
Connecting to EdgeDB instance 'edb' at localhost:10702...
edgedb error: SchemaDefinitionError: cannot alter object type 'default::LoanPayments' because this affects body expression of function 'func::getLateDays(id: std::uuid, date: cal::local_date)'
  Detail: cannot alter function 'func::getLateDays(id: std::uuid, date: cal::local_date)' because this affects expression of trigger 'update_loanAccruals' of object type 'default::LoanPayments'
edgedb error: cannot proceed until .esdl files are fixed
PS C:\Users\msavi\Documents\melfyn-v0\edb> edgedb migration create
Connecting to EdgeDB instance 'edb' at localhost:10702...
edgedb error: SchemaDefinitionError: cannot alter object type 'default::LoanPayments' because this affects body expression of function 'func::getOutstandingBalance(loan: default::Loan, loanPayment: OPTIONAL default::LoanPayments)'
  Detail: cannot alter function 'func::getOutstandingBalance(loan: default::Loan, loanPayment: OPTIONAL default::LoanPayments)' because this affects expression of trigger 'update_loanAccruals' of object type 'default::LoanPayments'
edgedb error: cannot proceed until .esdl files are fixed
PS C:\Users\msavi\Documents\melfyn-v0\edb> edgedb instance destroy -I "edb" --force
Unlinking \\?\C:\Users\msavi\Documents\melfyn-v0\edb
Instance edb is successfully deleted.
PS C:\Users\msavi\Documents\melfyn-v0\edb> edgedb project init
Found `edgedb.toml` in \\?\C:\Users\msavi\Documents\melfyn-v0\edb
Initializing project...
Specify the name of EdgeDB instance to use with this project [default: edb]:
> edb
Checking EdgeDB versions...
┌────────────────────────┬────────────────────────────────────────────────────────┐
│ Project directory      │ \\?\C:\Users\msavi\Documents\melfyn-v0\edb             │
│ Project config         │ \\?\C:\Users\msavi\Documents\melfyn-v0\edb\edgedb.toml │
│ Schema dir (non-empty) │ \\?\C:\Users\msavi\Documents\melfyn-v0\edb\dbschema    │
│ Installation method    │ WSL                                                    │
│ Version                │ 4.2+8778fc1                                            │
│ Instance name          │ edb                                                    │
└────────────────────────┴────────────────────────────────────────────────────────┘
Version 4.2+8778fc1 is already downloaded
Initializing EdgeDB instance...
Applying migrations...
Everything is up to date. Revision initial
Project initialized.
To connect to edb, run `edgedb`
PS C:\Users\msavi\Documents\melfyn-v0\edb> edgedb migration create
Connecting to EdgeDB instance 'edb' at localhost:10702...

```

**Comment by mkhlsavin** [comment]
After 5-6 attempts the schema will be compiled.

**Comment by aljazerzen** [comment]
Hmm, this is strange - migration generation process is supposed to be deterministic. If if gives an error in one run, it should also give the same error in all subsequent runs. Or ideally, generate migration correctly.

This does seem to be another issue, a bit related to this one. I'll file it separately when I get your schema.
